Milestone3 - Opshin Edge case Identification (#7)
Anastasia-Labs/opshin-audit: catalyst_docs/milestone3/milestone3-edgecases.md

# OpShin Edge Case Identification Report

**Project**: OpShin Audit  
**Milestone Title** : Milestone3 - Edge Case Identification  
**Submission Date** : 23-May-2025  
**Audit Team** : Anastasia Labs  
**Repository** : [Opshin Repository](https://github.com/OpShin/opshin)

## Executive Summary

This report documents edge cases in OpShin smart contract development identified through manual code review. It provides actionable strategies to resolve these issues that can be applied to future iterations of the language.

> ### Scope

Reviewer: ```suggestion
## Scope
```

We took the following approach to list the edge cases:

- Cross-referencing unit test files under `opshin/tests` against language specifications to uncover untested functionality.

- Incorporating documented edge cases from Milestone 2 to highlight high-risk scenarios lacking test coverage.

> # Edge cases

Reviewer: ```suggestion
## Edge cases
```

> ## 1. Untested builtin `len()`

Reviewer: ```suggestion
### 1. Untested builtin `len()`
```

Reviewer: And so on... we should have a consistent document structure (# Title should appear only once):

```
# OpShin Edge Case Analysis Report
## Summary
## Scope
## Edge cases
### Finding 1...
### Finding 2...
...
### Final remarks/notes/administation, etc.
```

1. Empty records are not included in the test cases, which leads to a lack of validation for subtyping behavior with empty records.

2. There are no test cases for records with fields in a different order.

3. There are no test cases for records that have the same fields but different constructor IDs.

### Recommendation:

We recommend to add similar tests for the above mentioned edge cases

1. Empty Records:

```
Empty = RecordType(Record("Empty", "Empty", 0, []))
```

2. Reversed field order:

```
A = RecordType(Record("A", "A", 0, [("foo", IntegerInstanceType), ("bar", ByteStringInstanceType)]))

A_reversed = RecordType(Record("A", "A", 0, [("bar", ByteStringInstanceType), ("foo", IntegerInstanceType)]))
```

3. `Records` having same fields but different constructor ids.

```
A = RecordType(Record("A", "A", 0, [("foo", IntegerInstanceType), ("bar", ByteStringInstanceType)]))

A_diff_constr = RecordType(Record("A", "A", 1, [("foo", IntegerInstanceType),("bar", ByteStringInstanceType)]))

```

## 6. Uncovered Scenarios in `Union` Type Behavior

- **File Reference**: `opshin/tests/test_Unions.py`
- **Edge Cases**:
  The following edge cases were not included in the test suite.

>   1.Duplicate entries in `Unions` give compiler errors, but duplicate entries in nested `Unions` don't.

Reviewer: Let's list the edge cases first and then proceed to the recommendations section. Similar to this:

Edge Cases:
1. ...
2. ...
3. ...

Recommendations:
1. ...
2. ...
3. ...

Duplicates should be identified based on `CONSTR_ID`, regardless of data field equivalence.

2. `Union`s with no entries or only one entry are not tested.
   Compiling the following code throws an unrelated error message: `'Name' object has no attribute 'elts'`.

```python
from opshin.prelude import *

@dataclass()
class A(PlutusData):
    x: int

def validator(a: Union[A]) -> None:
    assert isinstance(a, A)
```

### Recommendation:

The compiler should detect `Union`s containing only a single entry, and throw an explicit error.

3. `Dict` with `Union` type key, can't be accessed with a `Union` type which has the same entries but in a different order.

```python
from opshin.prelude import *

def validator(d: Dict[Union[int, bytes], int]) -> int:
    key: Union[bytes, int] = 0
    return d[key]
```

### Recommendation:

We recommend to sort the `Union` entries in an unambiguous manner within the `union_types()` function in `type_inference.py`.

## 7. Explicit testing for edge cases in bitmap and fractions

- **File Reference**: `opshin/tests/test_bitmap.py`,`opshin/tests/test_fractions.py`
- **Edge Case**:

> `test_fractions.py`: The existing sample cases exclude scenarios where the denominator is zero.

Reviewer: ```suggestion
`test_fractions.py`: The existing sample cases does not include scenarios where the denominator is zero.
```

> `test_bitmap.py`: Similar oversight with excluding empty bytestrings and negative values,these should also be tested to ensure robustness.

Reviewer: No, we should not try and figure out why the issue is present!
```suggestion
`test_bitmap.py`: Similarly there are no test cases covering empty bytestring inputs and negative input values (e.g. creating a bitmap with a negative length).
```

### Recommendation:

Add explicit test cases for

- Denominator = 0 in fractions (to check handling of division by zero).

> - Empty bytestrings and negative index values in bitmap operations.

Reviewer: ```suggestion
- Empty bytestrings and negative index/length values in bitmap operations.
```

## 8. Index type not tested for different types

- **File Reference**: `opshin/tests/test_stdlib.py`

- **Edge Case**:

> Attribute of listtype "index" is only tested for integers, missing other types which leads to the following edge case that wasn't tested.

Reviewer: ```suggestion
Attribute of listtype "index" is only tested for integers, missing other types, for example:
```

```python
from opshin.prelude import *

def validator(a: Anything, b: Anything) -> int:
   l: List[Anything] = [a, b]

   return l.index(b)
```

### Recommendation:

> In the `index` method, defined in `ListType.attribute()` in `type_impls.py`, change the check to `EqualsData(transform_output_map(itemType)(x), transform_output_map(itemType)(HeadList(xs)))`.

Reviewer: I don't think we should explicitly dictate code; we should just describe the gist of the recommended changes.

Author: changed to 
In the `index` method defined in `ListType.attribute()` in `type_impls.py`, update the check to accommodate other primitive types, and add corresponding test cases to ensure coverage.

## 9. No test cases for `CONSTR_ID` attribute.

- **Edge Case**:

> No test cases for analysing the behaviour of `CONSTR_ID` attribute.

Reviewer: ```suggestion
No test cases for validating the computation of `CONSTR_ID` attributes.
```

> The following edge cases are valid OpShin, but isn't consistent with how attributes are exposed of regular `Union`s (they must exist on each subtype), and can lead to unexpected runtime errors.

Reviewer: ```suggestion
The following edge cases are valid in OpShin, but isn't consistent with how attributes are used in other types (e.g. `Union`s).
```

```python
from opshin.prelude import *

def validator(l: List[Anything]) -> int:
    return l[0].CONSTR_ID
```

```python
from opshin.prelude import

def validator(u: Union[int, bytes]) -> int:

> return u.CONSTR_ID

Reviewer: ```suggestion
    return u.CONSTR_ID
```

```

### Recommendation:

1. Remove the `CONSTR_ID` attribute for `Anything`.

> 2. Don't expose the `CONSTR_ID` attribute of `Union`s which contain some non-`ConstrData` types.

Reviewer: ```suggestion
2. Do not expose the `CONSTR_ID` attribute of `Union`s which contain some non-`ConstrData` types.
```

> # Summary

Reviewer: There is already a summary section on top, this is definitely not needed. Please merge the contents into the one on top.

Author: I have merged it with the executive summary

This report identifies and analyzes 9 edge cases across the OpShin codebase and test suite.
Each case highlights a gap in either functionality coverage or robustness in the current implementation of the language.

For each edge case, we have proposed actionable recommendations, such as test coverage expansion,and improvements to type checking logic.
Implementing these changes will improve OpShin’s reliability, developer experience, and readiness for production smart contract development.

> # Validation by OpShin Team

Reviewer: ```suggestion
## Validation by OpShin Team
```

Reviewer: I don't think this section is needed at all.

Author: Removed
